skglm/solvers/gram.py

```diff
@@ -0,0 +1,235 @@
+import numpy as np
+from numba import njit
+from numpy.linalg import norm
+from celer.homotopy import _grp_converter
+
+from skglm.utils import BST, ST, BST_vec, ST_vec
+
+
+@njit
+def primal(alpha, r, w, weights):
+    p_obj = (r @ r) / (2 * len(r))
+    return p_obj + alpha * np.sum(np.abs(w * weights))
+
+
+@njit
+def primal_grp(alpha, norm_r2, r, w, grp_ptr, grp_indices, weights):
+    p_obj = norm_r2 / (2 * len(r))
+    for g in range(len(grp_ptr) - 1):
```

Comment:
in celer we needed to skip if weights[g] was np.inf because otherwise we ended up with 0 * np.inf which was nan

same in primal